tactics/tactics.ml

```diff
@@ -315,6 +315,36 @@ let convert_concl ~cast ~check ty k =
     end
   end
 
+let with_reverse_casts env env' sigma c =
+  let ev, l = EConstr.destEvar sigma c in
+  let () = match l with
+    | Nil | Default (_, Nil) -> ()
+    | _ -> assert false
+  in
+  let rec remake_instance acc k = function
+    | [] -> acc
+    | false :: rest -> remake_instance (SList.default acc) k rest
+    | true :: rest -> remake_instance (SList.cons (mkRel k) acc) (k-1) rest
+  in
+  let rec aux (k,bnd) env env' =
+    match env, env' with
+    | [], [] -> mkEvar (ev, remake_instance SList.empty k (List.rev bnd))
+    | d :: env, d' :: env' ->
+      (* is_section_variable: inserting letins for section variables
+         can produce spurious dependencies on them *)
+      if is_section_variable (Global.env()) (NamedDecl.get_id d)
+      || EConstr.eq_constr sigma (NamedDecl.get_type d) (NamedDecl.get_type d')
```

Comment:
And am I reading correctly that this keeps the trace for `change ... in H` but not `change ... in (value of H)`?

Reply:
yes